Add support for devices with multiple connected nodes

Lngramos opened the pull request:
## Context

The discovery code right now assumes that every device registered to a home/group maps to a single radiator unit. Whilst this might be true for homes with just one connected radiator or newer units that contain a built-in controller, it doesn't work for setups where radiators require an additional hub (Smartbox) to comm out to the API, as only the first node associated with a device is registered.

![Screenshot 2024-03-29 at 3 46 28 PM](https://github.com/duggan/homebridge-technotherm/assets/3799640/6eab4057-1569-4a2e-b03d-6b9a2b0ee872)

## Changes

This PR follows on the discussed in https://github.com/duggan/homebridge-technotherm/issues/28:

- Update the accessory discovery code to handle devices with multiple connected nodes
  - Accessory UUID generation is now seeded using the Node UID, previously was the Device UID. This will require that devices previously registered using the device ID to be manually removed from Homebridge cache, which is not high-impacting in this case, but please let me know if you disagree
- Check if nodes have been renamed against the accessories cache, and if so, update the renamed accessory within Homebridge

These changes should work for both your setup (@duggan) and mine. They are based on the following observations/assumptions:

- Devices represent the client that calls the Helki API
- A device can have multiple nodes associated with them
- A node represents a physical radiator/heater unit
- A node can only be connected to a single device at a time (at least for the units I have)

## Other

Changing the target temperature isn't working for me currently due to some API errors. I'm looking into this and raise a second PR to cover this separately.

Lngramos committed:
Add support for devices with multiple nodes

In the eye of the Helki API, devices represent the client calling their remote API.
Devices contain nodes. A node repesents a physical radiator/heater unit.
Some heaters ship with a built-in controller device capable of talking to the API
directly, in such cases, these devices contain a single node. Other heaters require
connecting to a central controller device, which means you end up with a Device containing multiple nodes.
This commit updates the device/node discovery logic to handle the latter scenario described above.
This in theory will work for users despite of the heater setup they have.
The accesory UUID is now seeded from the Node UID, previosuly was using the Device UID

Lngramos committed:
Add support for accessory renames

Check that if a node name has been modified and if so, notify Homebridge so the accessory is updated after initial discovery

Lngramos commented on .vscode/settings.json:
```diff
@@ -1,8 +1,8 @@
   "eslint.enable": true
-}
\ No newline at end of file
+}
```
These changes are automatically being made by the latest version of VS Code.

Happy to take these out, but thought might as well merge them, as I imagine you'll face the same. See https://code.visualstudio.com/updates/v1_85#_code-actions-on-save-and-auto

duggan replied:
Sounds ok to me – I don't use VS Code, these settings are actually part of the original homebridge template.

duggan commented:
@Lngramos that's an impressive patch, thanks! Hopefully will make this useful to many more people :) I've given it a quick spin and it looks like `node.uid` is not present in the results I get:

```
https://api.helki.com/api/v2/devs/0bfd0060XXXXXXX/mgr/nodes => {"nodes":[{"name":"","addr":2,"type":"htr","installed":true,"lost":false}]}
```

Re-including `device.dev_id` as an alternative seems to fix it though – can you confirm that will work for your setup? If so I'll merge this and publish a beta!

```diff
diff --git a/src/platform.ts b/src/platform.ts
index fa730d8..e58c29e 100644
--- a/src/platform.ts
+++ b/src/platform.ts
@@ -77,7 +77,7 @@ export class Technotherm implements DynamicPlatformPlugin {
           // Devices can have multiple nodes, register an accessory for each
           for (const node of nodes) {
             const accessoryName = node.name || device.name;
-            const accessoryUUID = this.api.hap.uuid.generate(node.uid); // Use dev_id to generate a UUID
+            const accessoryUUID = this.api.hap.uuid.generate(node.uid || device.dev_id); // Use dev_id to generate a UUID
             const existingAccessory = this.accessories.find(accessory => accessory.UUID === accessoryUUID);
 
             if (existingAccessory) {
```

Lngramos commented:
Here's what I receive on my end:

```
API GET request for https://api-evconfort.helki.com/api/v2/devs/e698XXXXXX/mgr/nodes => {"nodes":[{"name":"Living Room #1","addr":2,"parent":1,"level":1,"type":"acm","installed":true,"lost":false,"uid":"32122BXXXXXXXXXXXXXX"},{"name":"Living Room #2","addr":3,"parent":1,"level":1,"type":"acm","installed":true,"lost":false,"uid":"7AFD03XXXXXXXXXXXXXX"},{"name":"Guest Room","addr":4,"parent":1,"level":1,"type":"acm","installed":true,"lost":false,"uid":"20080AXXXXXXXXXXXXXX"},{"name":"Corridor","addr":5,"parent":1,"level":1,"type":"acm","installed":true,"lost":false,"uid":"23E28EXXXXXXXXXXXXXX"},{"name":"Bedroom","addr":6,"parent":2,"level":2,"type":"acm","installed":true,"lost":false,"uid":"881CE3XXXXXXXXXXXXXX"},{"name":"Office","addr":7,"parent":1,"level":1,"type":"acm","installed":true,"lost":false,"uid":"DC3333XXXXXXXXXXXXXX"}]}
```

I can confirm that it works on my end, but I had a feeling that `node.uid` might not be available on your end.

Let me amend the code to add the fallback to `device.dev_id` in that case.

Lngramos committed:
Fallback UUID generation using node.uid to device.dev_id

@duggan confirmed that node.uid is not defined on their end, so falling back to the original bahviour device.dev_id should handle this edge case whilst also maintaining backwards compatibility with the current implemention.

duggan commented:
@Lngramos beta's been published and I've verified it updates and runs ok for me :) `1.1.1-beta`.

In the unlikely event it doesn't work for you let me know, otherwise I'll publish it as a non-beta tomorrow! Thanks for the contribution, I hope others find it useful!

Lngramos commented:
Thanks! I can confirm `1.1.1-beta` works for me.

As I also mentioned before, the target temperature setting currently doesn't work for me. I haven't had a chance to properly investigate yet, but will take a gander hopefully sometime this week and hopefully have another PR coming your way soon.
